### ISSUE TITLE
Finer-grained response validation

### ISSUE THREAD

**Issue by TanukiSharp** [anttiviljami/openapi-backend#17]
Currently the output validation is performed per operation, as implemented in the `buildResponseValidatorForOperation` function. It ignores the status, and then create a `oneOf` schema. I think it could bring more value to take the status code into account.

I would like to suggest an improvement here.

The function `validateResponse` could take that status code as argument, in order to keep it framework independent and not having to grab it from the `res` argument. To this end, the function `getResponseValidatorForOperation` would need to take the status as argument too.

Therefore, probably requiring to change
```JS
public responseValidators: { [operationId: string]: Ajv.ValidateFunction };
```
to
```JS
// made multi-line for clarity purpose
public responseValidators: {
    [operationId: string]: {
        [statusCode: string]: Ajv.ValidateFunction
    }
};
```

Note that I can contribute by making a PR if that can help you.

Thank you for your consideration.

**Comment by anttiviljami** [comment]
Sounds fantastic! A PR would be great so we can continue the discussion about specifics. 

This could also solve the problem of the response oneOf error messages sometimes being vague and not too useful.

**Comment by anttiviljami** [comment]
pinging @akatsukle if you're also working on a PR regarding the structure of response validation. :)

**Comment by akatsukle** [comment]
I actually implemented the PR regarding #16 doing this, taking the statusCode as a parameter of validateReponseHeaders. 
I guess some of the logic in the PR could be reused, the part that matches the statusCode to a response. Its not really long nor hard, but factorizing the handling of `X` wildcard and the fallback to `default` could be nice.

**Comment by anttiviljami** [comment]
@akatsukle could you post your PR so me and @TanukiSharp can take a look?

**Comment by akatsukle** [comment]
Sure, [here is it](https://github.com/anttiviljami/openapi-backend/pull/18).
Sorry for not posting it earlier, I was not completely done with a few details.

The logic to match the correct code is not in its own function, but it should be fairly easy to extract ;).

**Comment by TanukiSharp** [comment]
@anttiviljami What about backward compatibility ? Technically speaking, if we want to be strict with SemVer, if I had a number (non-nullable) argument to a public method, you should make a 2.0.0 version. You probably don't want that.

I could make the `statusCode` as `number | undefined` and fallback to the old mechanism in case of `undefined`, but in such a case the implementation would get a bit trickier, so I would like to have your opinion about that.

Thanks.

**Comment by anttiviljami** [comment]
Yes the statusCode argument should definitely be optional. We need to preserve previous functionality

**Comment by anttiviljami** [comment]
> The logic to match the correct code is not in its own function, but it should be fairly easy to extract ;)

Brilliant idea! Let's extract the operation status code defaulting behaviour into its own method and use it across `validateResponseForOperation`, `validateResponseHeadersForOperation`, `mockResponseForOperation`. :+1:

**Comment by mt-sebastien-robert** [comment]
Hello there!

I guess you are busy working on different things. Just to know if you had time to review PRs #18 and #19 ?

Thanks.

**Comment by TanukiSharp** [comment]
Hi there!

I guess you are busy working on different things. Just to ask if you had a look at PRs #18 and #19 ?

Thanks.

**Comment by anttiviljami** [comment]
Hi @TanukiSharp ! Sorry, I've had a super busy week with work not related to openapi-backend. Checking your updates right now :)
